Add backup services initial (#25)
portworx/pxdocs-backup: content/use-px-backup/backup-stateful-applications/mysql.md

---
title: MySQL
description: 
keywords: 
weight: 4
hidesections: true
disableprevnext: true
---

This page provides instructions for users on how to use pre and post backups rules with PX-Backup to achieve application-consistent backups for MySQL in production.

> ## Background

Reviewer: You can do away with this heading and just jump directly into the concept information. 

```suggestion

```

> Information managed by MySQL server is stored in a location called the [data directory](https://dev.mysql.com/doc/refman/8.0/en/data-directory.html). Often, this data directory is located in the MySQL server filesystem at `/var/lib/mysql`. Within this location are the files and data that are vital for MySQL to persist information, and this is why it’s important to mount Kubernetes PersistentVolumeClaims (PVCs) to the data directory location that the MySQL image uses. In Kubernetes, the spec file volumeMount may look like this:

Reviewer: take a stab at active voice in this section and I can review again.

          claimName: mysql-data
```

Within the data directory, MySQL stores schema, table, logs, configuration and database data that corresponds to system, performance, and client data. Mounting a PersistentVolume enables PX-Backup to snapshot and back up MySQL data when needed.

MySQL has a tool called mysqldump which works well for taking backups specific to MySQL. Since PX-Backup can provide abstracted backup and restore for a polyglot of data services, we can replicate the [best practices of mysqldump](https://dev.mysql.com/doc/refman/5.7/en/mysqldump.html#option_mysqldump_add-locks), such as flushing and locking logs and tables using PX-Backup’s pre and post backup rules. Backups and rules can be used across multiple instances of MySQL and multiple types of clouds. This will ultimately simplify operations for devops teams for single or multi-cloud workflows.

For more information on how to run MySQL on Kubernetes, refer to [this article](https://portworx.com/mysql-kubernetes/ ).

## Installation

### Prerequisites

{{% content "shared/common-info.md" %}}

**NOTE:** MySQL pods must also be using the `app=mysql` label.

### Create rules for MySQL

> 

Reviewer: ```suggestion
Create rules for MySQL that will run both before and after the backup operation runs:

```

> #### Create a pre- backup rule for Cassandra

Reviewer: pre-exec?

```suggestion
#### Create a pre-exec backup rule for Cassandra
```

Before a backup of MySQL occurs, Portworx, Inc. recommends you to flush certain data to disk so that the backup remains consistent. Database tables and logs are examples of data that should be flushed. It is also important in MySQL to lock the tables so no new I/O transactions attempt to add records during the backup, or MySQL may also become inconsistent. 

To accomplish this, the [(`FLUSH TABLES WITH READ LOCK`)](https://dev.mysql.com/doc/refman/8.0/en/flush.html#flush-tables-with-read-lock) command is used in a pre-backup rule. The commandd will lock the tables with a global read lock and allow Portworx to take a application-consistent snapshot.

> **Note:** Set this rule to run in the background. This requires a `WAIT_CMD` to allow the rule to execute and exit properly.

Reviewer: ```suggestion
{{<info>}}
**NOTE:** Set this rule to run in the background. This requires a `WAIT_CMD` to allow the rule to execute and exit properly.
{{</info>}}
```

Create the following rule within the PX-Backup interface. Modify the username and password for how they are configured in your Kubernetes environment.

1. Navigate to **Settings** → **Rules** → **Add New**.
2. Add a name for your Rule.
3. Add the app label

	```text
	app=mysql
	```

4. Add the Action

	```text
	mysql --user=root --password=$MYSQL_ROOT_PASSWORD -Bse 'FLUSH TABLES WITH READ LOCK;system ${WAIT_CMD};'
	```

>  ![](/img/mysql-pre-rule.png)

Reviewer: ```suggestion
	 ![](/img/mysql-pre-rule.png)
```

> #### Create a post- backup rule for Cassandra

Reviewer: ```suggestion
#### Create a post-exec backup rule for Cassandra
```

PX-Backup performed a flush and lock on the data in MySQL before our backup. Now, PX-Backup must make sure to also run `UNLOCK` so the database releases the global read lock. It also may be a good idea to use [`FLUSH LOGS`](https://dev.mysql.com/doc/refman/8.0/en/flush.html#flush-logs). `FLUSH LOGS` will rotate logs and update the sequence number of the log. `FLUSH LOGS` is useful when users need a clear distinction between logs before and after a backup occurs. 

> **Note:** 

Reviewer: This needs to be wrapped in an info shortcode

```suggestion
{{<info>}}
**Note:** 
```

* Post backup rules are not allowed to run in the background, a WAIT_CMD is not needed.

> * Flushing logs is optional here but will add it to our post- backup rule for completeness.

Reviewer: ```suggestion
* Flushing logs is optional here but will add it to our post- backup rule for completeness.
{{</info>}}
```

Create the following rule within the PX-Backup interface. Modify the username and password for how they are configured in your Kubernetes environment.

1. Navigate to **Settings** → **Rules** → **Add New**.
2. Add a name for your Rule.

> 3. Add the app label

Reviewer: ```suggestion
3. Add the app label:
```

	```text
	app=mysql
	```

> 4. Add the Action

Reviewer: ```suggestion
4. Add the Action:
```

	```text
	mysql --user=root --password=$MYSQL_ROOT_PASSWORD -Bse 'FLUSH LOGS; UNLOCK TABLES;'
	```

>  ![](/img/mysql-post-rule.png)

Reviewer: ```suggestion
	 ![](/img/mysql-post-rule.png)
```

### Use the rules during backup of MySQL

> During the backup creation process, select the rules in the **pre-exec** and **post-exec** drop downs. An example of what this looks like is below:

Reviewer: ```suggestion
During the backup creation process, select the rules in the **pre-exec** and **post-exec** drop downs:
```

During the backup creation process, select the rules in the **pre-exec** and **post-exec** drop downs. An example of what this looks like is below:

![](/img/mysql-use-rules.png)

Once you've filled out the backup form, click **Create**

## Demo

> Watch this short demo of the above information.

Reviewer: ```suggestion

Watch this short demo of the above information.
```

---
title: MySQL
description: 
keywords: 
weight: 4
hidesections: true
disableprevnext: true
---

> This page provides instructions for users on how to use pre- and post- backups rules with PX-Backup to achieve application consistent backups for MySQL in production.

Reviewer: ```suggestion
This page provides instructions for users on how to use pre- and post- backups rules with PX-Backup to achieve application-consistent backups for MySQL in production.
```

## Background

> Information managed by MySQL server is stored in a location called the [data directory][2]. Often, this data directory is located in the MySQL server filesystem at `/var/lib/mysql`. Within this location are the files and data that are vital for MySQL to persist information, and this is why it’s important to mount Kubernetes PersistentVolumeClaims (PVCs) to the data directory location that the MySQL image uses. In Kubernetes, the spec file volumeMount may look like this:

Reviewer: I think we should replace all the footnotes with regular links.

```yaml
        volumeMounts:
        - name: mysql-data
          mountPath: /var/lib/mysql
      volumes:
      - name: mysql-data
        persistentVolumeClaim:
          claimName: mysql-data
```

Within the data directory, MySQL stores schema, table, logs, configuration and database data that corresponds to system, performance, and client data. Mounting a PersistentVolume enables PX-Backup to snapshot and back up MySQL data when needed.

MySQL has a tool called mysqldump which works well for taking backups specific to MySQL. Since PX-Backup can provide abstracted backup and restore for a polyglot of data services, we can replicate the [best practices of mysqldump][3], such as flushing and locking logs and tables into PX-Backup’s pre- and post- backup rules. PX-Backup rules and backups can be used across multiple instances of MySQL and multiple types of clouds. This will ultimately simplify operations for devops teams for single or multi-cloud workflows.

For more information on how to run MySQL on Kubernetes, read [this article][1].

## Installation

### Prerequisites

{{% content "use-px-backup/backup-stateful-applications/shared/common-info.md" %}}

{{<info>}}

> MySQL pods must also be using the `app=mysql` label.

Reviewer: ```suggestion
**NOTE:** MySQL pods must also be using the `app=mysql` label.
```

{{</info>}}

### Create rules for MySQL

#### Create a pre- backup rule for Cassandra

> Before a backup of MySQL occurs, it is recommended to flush certain data to disk so that the backup remains consistent. Database tables and logs are examples of data that should be flushed. It is also important in MySQL to lock the tables so no new I/O transactions attempt to add records during the backup, or MySQL may also become inconsistent. 

Reviewer: I'd do pretty much anything to avoid the phrase "it is recommended". Active voice prefers we specify an entity that is doing the action. 

```suggestion
Before a backup of MySQL occurs, Portworx, Inc. recommends you flush certain data to disk so that the backup remains consistent. Database tables and logs are examples of data that should be flushed. It is also important in MySQL to lock the tables so no new I/O transactions attempt to add records during the backup, or MySQL may also become inconsistent. 
```

To accomplish this, we can implement a `FLUSH TABLES WITH READ LOCK` command in our pre-backup rule. This will perform the following operations:

[(`FLUSH TABLES WITH READ LOCK`)][4] - Closes all open tables and locks all tables for all databases with a global read lock. This is a very convenient way to get backups if you have a file system such as Veritas or ZFS that can take snapshots in time. Use `UNLOCK TABLES` to release the lock.

Since PX-Backup performs a snapshot of the persistent volume in Kubernetes, this will accomplish what we are looking for.

{{<info>}}

> Note: Set this rule to run in the background, which requires a `WAIT_CMD` to allow the rule to execute and exit properly.

Reviewer: ```suggestion
**NOTE:** Set this rule to run in the background, which requires a `WAIT_CMD` to allow the rule to execute and exit properly.
```

{{</info>}}

Create the following rule within the PX-Backup interface. You may need to modify the username and password depending on how MySQL is configured in your Kubernetes environment.

1. Navigate to **Settings** → **Rules** → **Add New**.
2. Add a name for your Rule.
3. Add the app label

> `app=mysql`

Reviewer: same step formatting comment as the previous ones. Use triple backticks and tab this line in.

4. Add the Action

> ```bash

Reviewer: If this doesn't produce a **copy** button, I'd recommend switching these to \```text

mysql --user=root --password=$MYSQL_ROOT_PASSWORD -Bse 'FLUSH TABLES WITH READ LOCK;system ${WAIT_CMD};'
```

 ![](/img/mysql-pre-rule.png)

#### Create a post- backup rule for Cassandra

> Since we made sure to flush and lock data from MySQL before our backup, we must make sure to `UNLOCK` the database from the global read lock. MySQL documentation tells us that this is because global locks are not implicitly unlocked after the `FLUSH TABLES WITH READ LOCK` operation. It also may be a good idea to `FLUSH LOGS`, which “Closes and reopens any log file to which the server is writing” and updates the sequence number of the log. This may be required if users need a clear distinction between logs before and after a backup occurs. Flushing logs is optional here but will add it to our post- backup rule for completeness.

Reviewer: I assume this quote comes from the MySQL documentation. I almost always recommend we don't include any content from outside sources at all. If you do want to keep this, I think this would be a good use of a footnote attribution. 

```suggestion
Since we made sure to flush and lock data from MySQL before our backup, we must make sure to `UNLOCK` the database from the global read lock. MySQL documentation tells us that this is because global locks are not implicitly unlocked after the `FLUSH TABLES WITH READ LOCK` operation. It also may be a good idea to `FLUSH LOGS`, which “Closes and reopens any log file to which the server is writing” and updates the sequence number of the log. This may be required if users need a clear distinction between logs before and after a backup occurs. Flushing logs is optional here but will add it to our post- backup rule for completeness.
```

{{<info>}}

> Note: Post- backup rules are not allowed to run in the background, a WAIT_CMD is not needed.

Reviewer: ```suggestion
**NOTE:** Post- backup rules are not allowed to run in the background, a WAIT_CMD is not needed.
```

{{</info>}}

Create the following rule within the PX-Backup interface. You may need to modify the username and password depending on how MySQL is configured in your Kubernetes environment.

1. Navigate to **Settings** → **Rules** → **Add New**.
2. Add a name for your Rule.
3. Add the app label

> `app=mysql`

Reviewer: same formatting and indent comment as the others

---
title: MySQL
description: 
keywords: 
weight: 4
hidesections: true
disableprevnext: true
---

This page provides instructions for users on how to use pre and post backups rules with PX-Backup to achieve application-consistent backups for MySQL in production.

MySQL server manages information stored in a location called the [data directory](https://dev.mysql.com/doc/refman/8.0/en/data-directory.html). Often, the data directory is located in the MySQL server filesystem at `/var/lib/mysql`. MySQL stores data within this location that are vital for MySQL. You must make sure to mount Kubernetes PersistentVolumeClaims (PVCs) to the data directory location. In Kubernetes, the spec file `volumeMount` may look like this:

> ```yaml

Reviewer: is `volumeMounts` supposed to be offset from `volumes` like this?

Author: yes

          mountPath: /var/lib/mysql
      volumes:
      - name: mysql-data
        persistentVolumeClaim:
          claimName: mysql-data
```

Within the data directory, MySQL stores schema, table, logs, configuration and database data that corresponds to system, performance, and client data. Mounting a PersistentVolume enables PX-Backup to snapshot and back up MySQL data when needed.

MySQL has a tool called `mysqldump` which works well for taking backups specific to MySQL. Since PX-Backup can provide abstracted backup and restore for a polyglot of data services, we can replicate the [best practices of mysqldump](https://dev.mysql.com/doc/refman/5.7/en/mysqldump.html#option_mysqldump_add-locks), such as flushing and locking logs and tables using PX-Backup’s pre and post backup rules. Backups and rules can be used across multiple instances of MySQL and multiple types of clouds. This will ultimately simplify operations for devops teams for single or multi-cloud workflows.

For more information on how to run MySQL on Kubernetes, refer to [this article](https://portworx.com/mysql-kubernetes/ ).

## Installation

### Prerequisites

{{% content "shared/common-info.md" %}}

> **NOTE:** MySQL pods must also be using the `app=mysql` label.

Reviewer: ```suggestion
{{<info>}}
**NOTE:** MySQL pods must also be using the `app=mysql` label.
{{</info>}}
```

---
title: MySQL
description: 
keywords: 
weight: 4
hidesections: true
disableprevnext: true
---

This page provides instructions for users on how to use pre and post backups rules with PX-Backup to achieve application-consistent backups for MySQL in production.

MySQL server manages information stored in a location called the [data directory](https://dev.mysql.com/doc/refman/8.0/en/data-directory.html). Often, the data directory is located in the MySQL server filesystem at `/var/lib/mysql`. MySQL stores data within this location that are vital for MySQL. You must make sure to mount Kubernetes PersistentVolumeClaims (PVCs) to the data directory location. In Kubernetes, the spec file `volumeMount` may look like this:

> ```yaml

Reviewer: ```suggestion
```text
```

        volumeMounts:
        - name: mysql-data
          mountPath: /var/lib/mysql
      volumes:
      - name: mysql-data
        persistentVolumeClaim:

>           claimName: mysql-data

Reviewer: ```suggestion
          claimName: mysql-data
...
```

Create rules for MySQL that will run both before and after the backup operation runs:

#### Create a pre-exec backup rule for Cassandra

Before a backup of MySQL occurs, Portworx, Inc. recommends you to flush certain data to disk so that the backup remains consistent. Database tables and logs are examples of data that should be flushed. It is also important in MySQL to lock the tables so no new I/O transactions attempt to add records during the backup, or MySQL may also become inconsistent. 

To accomplish this, the [(`FLUSH TABLES WITH READ LOCK`)](https://dev.mysql.com/doc/refman/8.0/en/flush.html#flush-tables-with-read-lock) command is used in a pre-backup rule. The command will lock the tables with a global read lock and allow Portworx to take a application-consistent snapshot.

{{<info>}}
**NOTE:** Set this rule to run in the background. This requires a `WAIT_CMD` to allow the rule to execute and exit properly.
{{</info>}}

Create the following rule within the PX-Backup interface. Modify the username and password for how they are configured in your Kubernetes environment.

1. Navigate to **Settings** → **Rules** → **Add New**.
2. Add a name for your Rule.

> 3. Add the app label

Reviewer: ```suggestion
3. Add the following app label:
```

	```text
	app=mysql
	```

> 4. Add the Action

Reviewer: ```suggestion
4. Add the following action:
```

	```text
	mysql --user=root --password=$MYSQL_ROOT_PASSWORD -Bse 'FLUSH TABLES WITH READ LOCK;system ${WAIT_CMD};'
	```

	 ![](/img/mysql-pre-rule.png)

#### Create a post-exec backup rule for Cassandra

PX-Backup performed a flush and lock on the data in MySQL before our backup. Now, PX-Backup must make sure to also run `UNLOCK` so the database releases the global read lock. It also may be a good idea to use [`FLUSH LOGS`](https://dev.mysql.com/doc/refman/8.0/en/flush.html#flush-logs). `FLUSH LOGS` will rotate logs and update the sequence number of the log. `FLUSH LOGS` is useful when users need a clear distinction between logs before and after a backup occurs. 

{{<info>}}
**NOTE:** 

* Post backup rules are not allowed to run in the background, a WAIT_CMD is not needed.
* Flushing logs is optional here but will add it to our post- backup rule for completeness.
{{</info>}}

Create the following rule within the PX-Backup interface. Modify the username and password for how they are configured in your Kubernetes environment.

1. Navigate to **Settings** → **Rules** → **Add New**.
2. Add a name for your Rule.

> 3. Add the app label:

Reviewer: ```suggestion
3. Add the following app label:
```

	```text
	app=mysql
	```

> 4. Add the Action:

Reviewer: ```suggestion
4. Add the following action:
```
